Remove SHELL_CMD and fix signals for child procs.

monkburger opened the pull request:
On Ubuntu, it defaults /bin/sh to the dash shell, which has some kind of strange monitor mode, unlike bash. It's just better to directly call execve() on the logger program, just like Apache's piped logging. 

Also, we should unblock signals just before execve() so downstream processes can see SIGTERM etc.

pandax381 commented:
I've never heard of `wordexp()` before, but it is useful.

May I request the following changes?

- Do not set the flags `WRDE_NOCMD` because we need command substitution.
- I want you to use `execvp()` instead of `execve()` to search for binary file from `$PATH`.

Also, it seems redundant that `wordexp()` is called in `ngx_http_pipelog_set_log()` and `ngx_http_pipelog_command_exec()`. I would like the result of `wordexp()` to be stored at `ngx_http_pipelog_set_log()`. It seems like it would be better to add a member of `wordexp_t` to `ngx_http_pipelog_pim_t` and store it there.

monkburger commented:
@pandax381 right now I updated the PR to remove NOCMD and use execvp()

I'll work on moving wordexp() into set_log() over the weekend if I can get some additional time. Maybe it is a good idea to add wordexp_t to pim_t 

I added it in set_log because that is where some existing configuration settings are. It can probably be consolidated from pipelog_command_exec() into set_log but I wanted to make sure it was somewhat easy to follow.

monkburger commented:
@pandax381 I updated the PR. Let me know what you think.

pandax381 commented on ngx_http_pipelog_module.c:
```diff
@@ -2247,7 +2247,7 @@ ngx_http_pipelog_command_exec (ngx_str_t *command, ngx_fd_t rfd, ngx_cycle_t *cy
     char cmd[1024];
     ngx_fd_t fd;
-    wordexp_t  w;
+    ngx_http_pipelog_pim_t pim;
```
It seems to be an uninitialized variable. I would suggest changing the first argument of `ngx_http_pipelog_command_exec()` to `wordexp_t *` and passing `&pim[idx].w` on the caller side.

monkburger replied:
It's going to be a few weeks before I can get to this due to the holidays.

pandax381 commented:
@monkburger I added a commit because I want to merge them.

monkburger reviewed (commented):
Looks good to me!

pandax381 commented:
The process of restoring the signal mask was removed because it is no longer needed once all signals are unblocked.

monkburger commented:
Looks like the recent changes are crashing:

```
(gdb) frame 0
#0  0x00007fc87c81a246 in ngx_http_pipelog_command_exec (w=w@entry=0x5638ec24e368, rfd=<optimized out>, cycle=cycle@entry=0x5638ec241980) at ngx_http_pipelog_module/ngx_http_pipelog_module.c:2266
2266	        execvp(w->we_wordv[0], w->we_wordv);
(gdb) l
2261	                close(fd);
2262	            }
2263	        }
2264	        sigfillset(&mask);
2265	        sigprocmask(SIG_UNBLOCK, &mask, NULL);
2266	        execvp(w->we_wordv[0], w->we_wordv);
2267	        exit(1);
2268	    default:
2269	        break;
2270	    }
(gdb)

```
Looks like on my test server, w->we_wordv[0] is NULL (line 2266 in the above)
